File: src/pr_reviewer/db/migrate.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from pathlib import Path

from pr_reviewer.db.client import close_pool, connection

MIGRATIONS_DIRECTORY = Path(__file__).with_name("migrations")
MIGRATION_LOCK_ID = 41920260825


@dataclass(frozen=True)
class Migration:
    filename: str
    sql: str
    checksum: str


def get_missing_applied_migration_filenames(
    applied_filenames: list[str],
    available_filenames: list[str],
) -> list[str]:
    available = set(available_filenames)
    return [filename for filename in applied_filenames if filename not in available]


def load_migrations() -> list[Migration]:
    migrations: list[Migration] = []
    for path in sorted(MIGRATIONS_DIRECTORY.glob("*.sql")):
        sql = path.read_text(encoding="utf-8")
        checksum = hashlib.sha256(sql.encode("utf-8")).hexdigest()
        migrations.append(Migration(filename=path.name, sql=sql, checksum=checksum))
    return migrations


def migrate() -> None:
    migrations = load_migrations()

    with connection() as conn:
        conn.execute("select pg_advisory_lock(%s)", (MIGRATION_LOCK_ID,))
        try:
            conn.execute(
                """
                create table if not exists schema_migrations (
                  filename text primary key,
                  checksum text not null,
                  applied_at timestamptz not null default now()
                )
                """
            )
            conn.commit()

            cursor = conn.execute("select filename, checksum from schema_migrations")
            applied_rows = [dict(row) for row in cursor.fetchall()]
            applied_filenames = [str(row["filename"]) for row in applied_rows]
            available_filenames = [migration.filename for migration in migrations]
            missing = get_missing_applied_migration_filenames(
                applied_filenames,
                available_filenames,
            )
            if missing:
                raise RuntimeError(f"Applied migration files are missing: {', '.join(missing)}")

            applied = {str(row["filename"]): str(row["checksum"]) for row in applied_rows}

            for migration in migrations:
                applied_checksum = applied.get(migration.filename)
                if applied_checksum is not None:
                    if applied_checksum != migration.checksum:
                        raise RuntimeError(f"Migration checksum mismatch: {migration.filename}")
                    continue

                with conn.transaction():
                    conn.execute(migration.sql)
                    conn.execute(
                        "insert into schema_migrations (filename, checksum) values (%s, %s)",
                        (migration.filename, migration.checksum),
                    )
        finally:
            conn.execute("select pg_advisory_unlock(%s)", (MIGRATION_LOCK_ID,))
            conn.commit()


def main() -> None:
    try:
        migrate()
        print("Database migrations complete.")
    finally:
        close_pool()


if __name__ == "__main__":
    main()

File: packages/core/src/db/migrate.ts
import { createHash } from "node:crypto";
import { readdir, readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { db } from "./client";

const migrationsDirectory = join(
  dirname(fileURLToPath(import.meta.url)),
  "migrations",
);

type Migration = {
  filename: string;
  sql: string;
  checksum: string;
};

async function loadMigrations(): Promise<Migration[]> {
  const filenames = (await readdir(migrationsDirectory))
    .filter((filename) => filename.endsWith(".sql"))
    .sort();

  return Promise.all(
    filenames.map(async (filename) => {
      const sql = await readFile(join(migrationsDirectory, filename), "utf8");

      return {
        filename,
        sql,
        checksum: createHash("sha256").update(sql).digest("hex"),
      };
    }),
  );
}

export async function migrate(): Promise<void> {
  const migrations = await loadMigrations();
  const client = await db.connect();

  try {
    await client.query("select pg_advisory_lock(41920260825)");
    await client.query(`
      create table if not exists schema_migrations (
        filename text primary key,
        checksum text not null,
        applied_at timestamptz not null default now()
      )
    `);

    for (const migration of migrations) {
      const applied = await client.query<{ checksum: string }>(
        "select checksum from schema_migrations where filename = $1",
        [migration.filename],
      );

      if (applied.rowCount === 1) {
        if (applied.rows[0]?.checksum !== migration.checksum) {
          throw new Error(`Migration checksum mismatch: ${migration.filename}`);
        }
        continue;
      }

      await client.query("begin");
      try {
        await client.query(migration.sql);
        await client.query(
          "insert into schema_migrations (filename, checksum) values ($1, $2)",
          [migration.filename, migration.checksum],
        );
        await client.query("commit");
      } catch (error) {
        await client.query("rollback");
        throw error;
      }
    }
  } finally {
    await client.query("select pg_advisory_unlock(41920260825)");
    client.release();
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  migrate()
    .then(() => {
      process.stdout.write("Database migrations complete.\n");
    })
    .finally(() => db.end());
}
